### ISSUE TITLE
Would Three consider a prototype.reset() function for its value classes (Vectors, Colors, etc)?

### ISSUE THREAD

**Issue by drcmda** [mrdoob/three.js#21209]
Being able to reset is a critical feature in deterministic design systems. For instance the DOM can do this:

```jsx
const a = document.createElement("div")
a.style.color // -> ""

a.style.color = "red"
a.style.color // -> "red"

a.style.color = null
a.style.color // -> ""
```

This is why we can simply delete a style from Chrome Devtools and it goes back to its default.

or in React we can write:

```jsx
<meshBasicMaterial color="red" />
```

Now we could simply remove the color, save the file, and hot module reload will tell the reconciler that "color" does not exist any longer on the same element:

```jsx
<meshBasicMaterial />
```

**But**, since there is no way to know the default, it is stuck on red. I discussed this with Dan Abramov, and he was surprised that Three doesn't have this. All other platforms seem to be able to do this, i asked around: Dom, Android, Ios, Konva, Ink, etc.

It would be a simple addition for all value types, so classes that have set/clone/copy. It could be a reset method for instance that puts the item back to the state it had:

```jsx
const color = new THREE.Color()
color.set("red")
color.reset()
color.getHexString() // -> "ffffff"
```

Or a defaults collection on the prototype:

```jsx
Color.prototype.defaultArguments = [0, 0, 0]
// ...
color.set(...color.defaultArguments)
```

This is not something that would only help React, it would also be relevant to any design system that is driven by Three, visual editors, IDEs and so on.

**Comment by donmccurdy** [comment]
Note that the default value of an object may depend on its context; if `mesh.up.reset()` changes its value to `[0,0,0]`, is that what you're expecting?

An option that is available now would be something like:

```js
const DEFAULT_COLOR = new THREE.Color();

// ...

material.color.copy(DEFAULT_COLOR);
```

**Comment by drcmda** [comment]
In a way Three already has it, for instance: https://github.com/mrdoob/three.js/blob/master/src/math/Color.js#L606-L609 but it would have to be generic so that you can know defaults without knowing the class or its keys, in this case: r, g and b.

Cloning would mean that the design system has to practically store and double every class systemwide itself, and with dangerous repercussions since clone contains the full object, not the arguments you construct it with. This is similar to a sphere, which is a radius, and segments, but a clone contains vertices and the full buffer.

>  if mesh.up.reset() changes its value to [0,0,0], is that what you're expecting?

I tried to find "up" on THREE.Mesh, but could not find it. Which class is it?

**Comment by donmccurdy** [comment]
It's a Vector3 that defaults to [0,1,0] — https://threejs.org/docs/#api/en/core/Object3D.up

To clarify, are you asking for a way to restore an object to the _default_ state for its type, or to the state in which it was constructed?

**Comment by drcmda** [comment]
> To clarify, are you asking for a way to restore an object to the default state for its type, or to the state in which it was constructed?

Only to the default state.

if a user does

const o = new Object3D()
o.up.set(0, 0, 1)
o.up.reset()

Then it must be 0/0/0 since reset should act purely/deterministically without context.

**Comment by mrdoob** [comment]
`Matrix3` and `Matrix4` have `identity()` but I don't `color.identity()` doesn't sound good... ?

Also, how many classes would this apply to?

**Comment by drcmda** [comment]
Oh, right, this is exactly what is needed. Color.identity sounds perfect! It could be added to all classes that are under the Math section. Had a quick look what their defaults are:

```jsx
  [THREE.Box2, [new THREE.Vector2(+Infinity, +Infinity), new THREE.Vector2(-Infinity, -Infinity)]],
  [THREE.Box3, [new THREE.Vector2(+Infinity, +Infinity), new THREE.Vector2(-Infinity, -Infinity)]],
  [THREE.Color, ["white"]],
  [THREE.Cylindrical, [1, 0, 0]],
  [THREE.Euler, [0, 0, 0, 'XYZ']],
  [THREE.Vector2, [0, 0]],
  [THREE.Vector3, [0, 0, 0]],
  [THREE.Vector4, [0, 0, 0, 1]],
  [THREE.Matrix3, [1, 0, 0, 0, 1, 0, 0, 0, 1]],
  [THREE.Matrix4, [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1]],
  [THREE.Quaternion, [0, 0, 0, 1]],
  [THREE.Sphere, [new THREE.Vector3(), -1]],
  [THREE.Ray, [new THREE.Vector3(), new THREE.Vector3(0, 0, -1)]],
  [THREE.Spherical, [1, 0, 0]],
  [THREE.Triangle, [new THREE.Vector3(), new THREE.Vector3(), new THREE.Vector3()]],
```

**Comment by mrdoob** [comment]
Curious to know what @WestLangley thinks about this...

**Comment by Mugen87** [comment]
> but color.identity() doesn't sound good... ?

Identity matrix makes sense. I've never heard of identity color and I'm afraid this semantic sounds quite uncommon. Same for all other math classes.

**Comment by Mugen87** [comment]
To me, there is no obvious way to define a "default state" for math entities. For example the semantic of a vector mainly depends on its usage. A direction vector (which always has unit length) is something different than a vector defining a position in 3D space. There is no way to define a reset value appropriate for all possible use cases.

**Comment by drcmda** [comment]
imo for vestigial reasons and because "identity" is already in use the benefits outweigh the disadvantages. otherwise you'd have another breaking change, or duplicated api names. for a design system resetting nodes without deeper knowledge of them is an invaluable feature.

> There is no way to define a reset value appropriate for all possible use cases.

a vector is always 0/0/0, simply because if you do new THREE.Vector3() no matter the context these are the default values, its "identity". as i demonstrated, in special cases it would be quite easy to reset to contextual values, like up being 0/1/0.

**Comment by gaearon** [comment]
Just to be clear, the problem we're trying to solve is not just resetting a whole node, but resetting a _particular property_ to its default value. Just like in the DOM, `.property = null` lets us get rid of any behaviors caused by `.property` being present.

**Comment by drcmda** [comment]
> Just like in the DOM, .property = null lets us get rid of any behaviors caused by .property being present.

One small explanation regarding this, in reacts `<mesh position={...} />`, position would refer to the THREE.Vector3 class. most props are just threes regular value classes. the others would already behave as you mention, for instance deleting children will wipe out sub-tree. Likewise, in a design system, IDE, or dev tools, it could behave in the same manner.

**Comment by gaearon** [comment]
The key distinction I want to emphasize is that we don't want to mutate the *value*. E.g. if `mesh.color` is an instance of a `Color`, in my mental model what we need is **not** `mesh.color.reset()`. Because that mutates the instance of the color. Instead, we want to get the `mesh` itself into the state it was before we assigned its `color`. Does this make sense?

E.g. if two meshes share the same instance of a Color, "resetting" that color would reset both. That is _not_ what a change from `<mesh color={red} />` to `<mesh />` represents. What it represents is that "this mesh no longer has this color", not "this mesh's color instance needs to be mutated".

**Comment by gaearon** [comment]
If we can't get this for some reason, I don't think the addition of "reset" functions on the values themselves would help us. Because indeed, there is no single "default value" for each type. There is a default value only for _each property_ where this type is being used, on particular objects. So our next preferred workaround would have to be to snapshot default values of each property when an object is created, so that we can revert values to the default ones when needed.

**Comment by drcmda** [comment]
Imo it would be exactly that. Mesh has it's own color, and if you pass a foreign color in, it will copy it.

```jsx
<mesh color="red" /> 
// -> mesh.color.set("red")

const red = new THREE.Color("red")
<mesh color={red} /> 
// -> mesh.color.copy(red)
```

**Comment by gaearon** [comment]
When `<mesh />` is created without a `color`, what is `mesh.color` equal to?

**Comment by drcmda** [comment]
the default of THREE.Color, which is white.

**Comment by gaearon** [comment]
Here's my problem with the `reset()` proposal.

Consider this example:

```js
const red = new THREE.Color("red")

function MyComponent({ step }) {
  if (step === 0) {
    return (
      <>
        <material color={red} />
        <material color={red} />
      </>
    )
  } else {
    return (
      <>
        <material />
        <material color={red} />
      </>
    )
  }
}
```

When `step` is `0`, both colors are red. When `step` is `1`, one of the materials loses color, but the second one doesn't.

With the `reset` proposal, I worry that we'll do `material1.color.reset()`, which — since it shares the `Color` instance with `material2`, would lead to the second `material` losing color.

Instead, conceptually, what we want to say is really `material1.color = null`, or, if that doesn't work, something like `material1.color = Material.DEFAULTS.color`.

**Comment by drcmda** [comment]
it won't share the instance. :-) it calls the local material.set. in threejs objects already have all the props they need, they aren't additive like on the dom. i guess in vanilla three you could actually exchange it, but r3f for certain doesn't. it doesn't even know what a color is. it just detects a class with a "set" method on it. the rest is handled by typescript making it safe.

**Comment by gaearon** [comment]
My other problem with the `reset` proposal is that there is no knowledge encoded in the value which one is the "default" one. For example, consider two hypothetical properties:

```js
material.speed // a vector
material.position // a vector
```

(I know these aren't real, but there's probably similar situations)

Here, both are vectors, but they may start differently. E.g. position might be a Zero vector, while speed may be a Unit vector. The value type (Vector) does not have the necessary knowledge to know what it should "reset to".

The only thing that has this knowledge is the thing that assigned the default value. The material constructor, in this example.

**Comment by drcmda** [comment]
that's what @Mugen87 said. imo that's the users concern. if they add and then delete values -> default no matter what. but for convenience the value class *could* set a contextual default. like that up vector above.

**Comment by gaearon** [comment]
>if they add and then delete values -> default no matter what.

I understand that this would make sense for a `reset()` function to do that. However, that `reset()` function would not be useful to a React renderer because that's not the desired semantic. The desired semantic from the React perspective is to reset _to the same value that we had initially_, not to some abstract "default value". So this proposal would make the behavior just as confusing, if not more, than it is now.

**Comment by cgauld** [comment]
Hi! I've been discussing this with @drcmda over on react-three-fiber :-)

I agree that this concept sits best at the property level, rather than at the type level. A given Vector3 may have a default value of [0, 0, 0] in one situation (e.g. for `position`) but [1, 1, 1] in another (e.g. `scale`). And since Vector3 is a reference value then we can't rely on one property having a unique Vector3 to reset without affecting other uses of it.

One thought I had was to use `Object.defineProperty` to provide getters and setters of the properties such that setting `undefined` on a property would reset it to its default value. Like this:
```
let myObject = new THREE.Mesh();
console.log(myObject.position); // outputs Vector3 with values 0, 0, 0

myObject.position.set(1, 2, 3);
console.log(myObject.position); // outputs Vector3 with values 1, 2, 3

myObject.position = undefined;
console.log(myObject.position); // outputs Vector3 with values 0, 0, 0
```
It reads a bit strange in this example but there is precedence for this in other paradigms, especially where getter/setter functions are used rather than direct property access.

**Comment by looeee** [comment]
> One thought I had was to use Object.defineProperty to provide getters and setters of the properties

I think getts and setters have too high a performance cost to be used for frequently accessed properties like position and scale.

**Comment by Mugen87** [comment]
This issue reminds me at #17845. 

TBH, it looks to me that the OP tries to fix a React issue on `three` level. I do not vote to change anything the library.

**Comment by drcmda** [comment]
this is not react related. all i would like is the ability to put an object back to its initial state. where this used, react, devtools, and so on, i think that doesn't matter much. essentially what matrix can already do with identity.

**Comment by gaearon** [comment]
I wouldn’t say the issue is on the React level. We can work around it on the React level, but the issue is just in the difference of capability between the models.

Imagine making a 3D editor with Three. This editor would allow to manipulate a Three tree in a graphical way, a la Photoshop. When you select an object, you can modify its properties in some tool bar.

If you were building such an editor, you would run into exactly the same problem. After you edit some value you’d have no way to get it back to the default aside from literally “undo”. Unless you actually know what the default is.

Some editors provide the ability to reset changes to the individual values. Others highlight the default value in the list so that you can which one you can choose to undo the changes. Yet others don’t provide this functionality at all.

I think it would be useful to expose the default values used by the constructors by some convention. So that either the React use case or the editor use case can introspect them.

If it’s not possible then the correct workaround on the React side would be to snapshot properties at creation — so that we may copy them over when deleting.

The proposed `reset` method, while it correctly solves a _subset_ of these cases, is not the correct solution for the React model because it would not work correctly for any case where the initial value differs from the type’s “default” value.

**Comment by cgauld** [comment]
> > One thought I had was to use Object.defineProperty to provide getters and setters of the properties
> 
> I think getters and setters have too high a performance cost to be used for frequently accessed properties like position and scale.

A very good point @looeee :-) I think in this case we could potentially get away with only using a setter, and leaving property access as normal. Most of the time object property setting is already by a setter on the value object anyway. So the cost would look like this:
```
// Regular property get, normal cost
console.log(myObject.position)

// Setting on value object, normal cost
myObject.position.set(1, 2, 3);

// Setting property itself, new cost since myObject.position setter is now function that checks for undefined
myObject.position = new Vector3(1, 2, 3);
```
The cost would be more pronounced for non-object properties:
```
// Regular property get, normal cost
console.log(myObject.visible);

// Setting value, new cost
myObject.visible = false
```
I'm not necessarily advocating this approach but thought it was an interesting way to address the problem 😉

**Comment by mrdoob** [comment]
TL;DR?

**Comment by cgauld** [comment]
I can have a go :-) It would be super useful if there was a way to reset three object properties to default values. React-three-fiber has a specific issue where this woud help but there are other use cases too. Options:
 - Implement `reset` functions on objects (e.g. Vector3) that return them to their initial value. Helps in some circumstances but not for pass-by-value properties (e.g. `visible`).
 - Use Object.defineProperty to add a settor function to properties such that assigning `undefined` returns the property to the default value (e.g. myObject.position = undefined returns it to 0, 0, 0). Incurs a performance regression since settors are slower than direct assignment.
 - Expose the defaults as static properties, e.g. Object3D.DEFAULTS.position

Let me know if I've missed anything!

**Comment by mrdoob** [comment]
Thank you!

As @donmccurdy said... Couldn't react-three-fiber create internal defaults like this?

```js
const DEFAULT_OBJECT3D = new Object3D();
```

And then do:

```js
mesh.position.copy( DEFAULT_OBJECT3D.position );
```

**Comment by looeee** [comment]
This would be a good time to use getters since resetting things is unlikely to happen every frame, so you could avoid creating defaults until they are used.

``` js
let defaultObject3D = null;
let defaultMesh = null;
let defaultMeshStandardMaterial = null;
...

const DEFAULTS = {
	get DEFAULT_OBJECT3D () {
		if (!defaultObject3d ){
   			defaultObject3d = new Object3D();
		} 
		return defaultObject3d;
	},
	get DEFAULT_MESH () {...},
	get DEFAULT_MESH_STANDARD_MATERIAL () {...},
	...	
}
```

Then to use it:

``` js
function resetObject3D(object3d) {
	object3d.copy(DEFAULTS.DEFAULT_OBJECT3D);
}

function resetMesh(mesh) {
	mesh.copy(DEFAULTS.DEFAULT_MESH);
}

...
```

...that's the basic idea, although I'm sure there are more elegant ways of writing this. This is what I'd do if I was creating a scene editor or similar. It's a fair argument to say this should be done at the application level.

On the other hand, it would be simple to add this functionality directly to three.js, so we'd have `Object3D.reset`, `Mesh.reset`, `MeshBasicMaterial.reset`, `Color.reset`, and so. 

It doesn't make sense to add this everywhere though - as others have pointed out, `Vector<x>.reset` has complications, and there's already `Matrix<x>.identity`. Perhaps we could limit it only to high-level components - `Scene`, `Object3D`, `Mesh`, materials, and so on? Then you could do `mesh.reset` but not `mesh.position.reset`.

**Comment by pailhead** [comment]
May i try a `tl:dr;`?

All three js works like this:
```
const vec3 = new Vector3() // 0,0,0
vec3.set(1,2,3) // 1,2,3
```
Often, if not always, we have a shorthand for this:
```
const vec123 = new Vector(1,2,3) // 1,2,3
```
What's being suggested here is:

```
someVector3.reset() // 6,6,6 -> 0,0,0
```

The problem is the shorthand. because you can instantiate a vector with some value (or a color or whatever) `reset` suddenly becomes impossible. 

```
const vec = new Vector3()
vec.set(1,2,3)
vec.set(0,0,0) // :)
vec.reset()    // :(
```

I think the solution could be:

```
const mesh = new Mesh()
console.log('by DEFAULT, three's Mesh.up is:', mesh.up)
```
(write this on paper) and then:
```
Mesh.prototype.reset = function(){
  this.speed = DEFAULT_SPEED
  this.up.set(0,1,0) //from paper
  this.material = new MeshBasicMaterial({color: getRandomColor()}) //this is essentially what happens
  this.favoriteSpaceQuest = `IV`
  this.position.reset() //because this is how EVERY mesh starts, and it's easier to write than set(0,0,0)
  this.scale.copy(DEFAULT_SCALE) //gotcha, this doesnt start at 0,0,0, so you can't use reset
}
```

So for the final `tl;dr` - i think some people may find it useful to `set` various objects found in three, to **the state they would start off if instantiated with no arguments**.

**Comment by pailhead** [comment]
My thoughts are - just because a color could (and should) be resettable to white, it doesn't mean that a `Mesh` **has** to reset it's up vector, just because it's a vector and has a reset method... It can "reset" it, by copying some default value, other than 0,0,0.

**Comment by pailhead** [comment]
>If it’s not possible then the correct workaround on the React side would be to snapshot properties at creation — so that we may copy them over when deleting.

@gaearon 

Can you elaborate on this? I'm confused by the plural of properties. Are you describing how to do this without hardcoding the default properties for each object?

```
const DEFAULT_OBJECT3D = new Object3D();
Object3D.prototype.reset = ()=>{ this.copy(DEFAULT_OBJECT3D } 
```

**Comment by pailhead** [comment]
How about:
```
const thingsNeededForReact  = [Object3D, Color, Vector2, Vector3, Matrix3, ...moar_stuff]
thingsNeededForReact.forEach(thing=>thing.prototype.reset = (()=>{
  const default = new thing()
  return ()=>{ this.copy(default) }
})())
```
Or something along those lines?

**Comment by mrdoob** [comment]
Sorry, I don't want this conversation to drag on.

The solution is to do something like this:

```js
const DEFAULTS = {
	'Color': new Color(),
	'Object3D': new Object3D(),
	'Vector3': new Vector3()
};
```

And then doing:

```js
mesh.position.copy( DEFAULTS[ 'Object3D' ].position );
```

Considering this is a specific `react-three-fiber` issue I think the code should sit there.
